### DISCUSSION TITLE
Observing throttle usage

### DISCUSSION THREAD

**Discussion by derekargueta** [kickstarter/rack-attack#653] [Questions (Q&A)]
Is there a way to instrument how much of the throttle quotas get used? As some kind of histogram, i.e. what the 50th percentile usage was, to inform whether the limits could use a little tightening. I'd imagine this getting captured when the period resets.

**Comment by santib** [comment]
Hi @derekargueta, I moved the issue to a discussion if that's fine.

>how much of the throttle quotas get used?

Let me see if I understand your question. Given a throttle setup, you want to know for each discriminator how close it was to getting throttled?

For example if you have
```ruby
Rack::Attack.throttle("requests by ip", limit: 5, period: 2) do |request|
  request.ip
end
```

You want to know for each period (2 secs) and each discriminator (IP) how many requests were made so to know how close they were to the limit (5 reqs) ?

I don't think there is any kind of instrumentation for that, [only when a request if matched we notify it](https://github.com/rack/rack-attack/blob/main/lib/rack/attack/throttle.rb#L46).

That said, you might be able to achieve your goal in some other ways. For example:
1. Using the [Tracks feature](https://github.com/rack/rack-attack/tree/main?tab=readme-ov-file#about-tracks)?
2. Maybe writing a custom cache store that decorates the one you are using, with logic to track every increment to the cache store?
